### PR TITLE
HPCC-32244 Add custom views to Metrics 

### PR DESCRIPTION
**Pull request by GordonSmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [x] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/src/src-react/hooks/metrics.ts [hunk 1]
```diff
@@ -2,96 +2,218 @@ import * as React from "react";
...
-const options: MetricsOptions = { ...defaults };
+// const _defaultUserViews = JSON.stringify({ viewId: _viewId.value, views: _views.value });
 
-function checkLayout(options: MetricsOptions): boolean {
-    if (options?.layout && !options?.layout?.["main"]) {
-        delete options.layout;
-    }
-    return !!options?.layout;
+export function clone<T>(_: T): T {
+    return JSON.parse(JSON.stringify(_));
```
Comment (jeclrsg) [inline esp/src/src-react/hooks/metrics.ts:102]: there's a relatively recent "structuredClone" global method in the JS API https://caniuse.com/?search=structuredClone

File: esp/src/src-react/hooks/metrics.ts [hunk 2]
```diff
@@ -2,96 +2,218 @@ import * as React from "react";
...
+}
+
+// function checkLayout(options: MetricsView): boolean {
```
Comment (jeclrsg) [inline esp/src/src-react/hooks/metrics.ts:105]: a bit of commented out code here and that _defaultUserViews a few lines up

File: esp/src/src-react/components/controls/Grid.tsx [hunk 3]
```diff
@@ -299,7 +305,32 @@ const FluentStoreGrid: React.FunctionComponent<FluentStoreGridProps> = ({
...
     }, [columnWidths]);
 
-    return <div data-is-scrollable="true" style={{ overflow: "auto", height: "100%" }}>
+    /*  Monitor Scroll Events (hack)
+
+        Essentially we are setting the scrollElement of the DetailsList to the div that contains the DetailsList (rather than a scrollable pane host).
+        See:  https://github.com/microsoft/fluentui/blob/55d3a31042e8972ea373841bef616c68e6ab69f9/packages/react/src/components/List/List.tsx#L355-L369
+
+        Note: Not sure if `_onScroll` call is needed, but excluding for now as it seems to work without it and is more performant.
+    */
```
Comment (jeclrsg) [inline esp/src/src-react/components/controls/Grid.tsx:314]: didn't this get removed a few weeks back in favor of that data-is-scrollable attribute?

File: esp/src/src-react/components/Metrics.tsx [hunk 4]
```diff
@@ -610,31 +625,28 @@ export const Metrics: React.FunctionComponent<MetricsProps> = ({
...
-        scopesTable
-            .metrics(metrics, options, scopeFilter, matchCase)
-            .render(() => {
-                updateScopesTable(selectedMetrics);
-            })
-            ;
+    }, []);
 
-    }, [matchCase, metrics, options, scopeFilter, scopesTable, selectedMetrics, updateScopesTable]);
+    console.log("View ID", viewId, view.scopeTypes);
```
Comment (jeclrsg) [inline esp/src/src-react/components/Metrics.tsx:640]: console.log here

File: esp/src/src-react/components/MetricsOptions.tsx [hunk 5]
```diff
@@ -1,121 +1,280 @@
...
-                    }} />
-                    <TextField label={nlsHPCC.EdgeLabel} value={options.edgeTpl} multiline autoAdjustHeight onChange={(evt, newValue) => {
-                        setOptions({ ...options, edgeTpl: newValue });
+    const onAddLabel = React.useCallback((label: string) => {
+        if (label) {
+            addView(label, { ...view, ...dirtyView });
+        }
+    }, [addView, dirtyView, view]);
+
+    console.log("dirtyView.scopeTypes", viewId, view.scopeTypes);
```
Comment (jeclrsg) [inline esp/src/src-react/components/MetricsOptions.tsx:158]: console.log here

**Review by jeclrsg — CHANGES_REQUESTED** [review]

File: esp/src/src-react/hooks/metrics.ts (hunk 2 above)
Comment (GordonSmith) [reply esp/src/src-react/hooks/metrics.ts:105]: Actually want to call this...

File: esp/src/src-react/hooks/metrics.ts (hunk 1 above)
Comment (GordonSmith) [reply esp/src/src-react/hooks/metrics.ts:100]: Good to know - I think I will leave this code alone for now until I have had a chance to see it in action.

**Review by jeclrsg — APPROVED** [review]

**Comment by GordonSmith** [comment]
@ghalliday - don't pull yet, I am re-testing and have found a potential issue...

**Pull request merged**